# This folder contains datasets generated by the repository.

## Functional annotation of target proteins  

+ [`human_druggable_genome_protein_class.tsv`](human_druggable_genome_protein_class.tsv) contains annotations of 5,197 proteins in human druggable genome, along with their functional classes and subclasses. Altogether there are 7 classes and 17 subclasses in total.

## Function similarity of target pairs

+ Function similarity of 70,876 target pairs was computed by: 
  + [similarity of Reactome annotation](function_similarity/descriptor_all_select_features_mc_0.85_function_similarity_reactome.tsv)
  + [similarity of GO annotation](function_similarity/descriptor_all_select_features_mc_0.85_function_similarity_go.tsv) 
Target pairs were then grouped into ones with common function (functionn similarity > 0) and ones without common function (functionn similarity = 0). The grouped result can be found in [this file](function_similarity/descriptor_all_select_features_mc_0.85_function_similarity_group.tsv), along with the selected feature similarity of target pairs. 

## Differential expression of selected target genes 

+ [Differential expression values of 28,051 genes](target_expression/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_uniprot_median_tpm_adjust.tsv) were computed.

+ Fifteen adverse events of interest were manually [mapped to GTEx tissues](target_expression/adverse_event_tissue_map.tsv).

+ Differential expression values in the matched tissue were extracted for the selected target genes. [Extracted results of 15 adverse events](target_expression/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_uniprot_median_tpm_adjust_all_adverse_event_select_targets_compare_de.tsv) were combined. Differential expression values were compared between selected target genes and all measured genes for each adverse event by [Wilcoxon test](target_expression/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_uniprot_median_tpm_adjust_all_adverse_event_select_targets_compare_pv.tsv).  

## Disease connection of selected target genes 

+ [32,279 gene-disease connections with direct evidence](target_disease/CTD_genes_diseases_direct_evidence_data.tsv) were extracted from CTD. In total, [8,679 genes](target_disease/CTD_genes_diseases_direct_evidence_genes.txt) are involved in these connections.

+ Thirteen adverse events of interest were manually [mapped to seven disease groups and sets of key words](target_disease/adverse_event_disease_term_map.tsv) for disease term matching. By search of key words, [677 disease terms](target_disease/CTD_genes_diseases_direct_evidence_diseases.tsv) were identified for the seven disease groups. [75 disease terms](target_disease/CTD_genes_diseases_direct_evidence_diseases_remove.tsv) were manually removed due to their irrelevance with the matched disease groups. [The remaining 602 terms](target_disease/CTD_genes_diseases_direct_evidence_diseases_filtered.tsv) can be mapped to [7,287 gene-disease connections with direct evidence](target_disease/CTD_genes_diseases_direct_evidence_diseases_filtered_connection_map.tsv).

+ Using processed data above, [selected target genes of 13 adverse events](target_disease/descriptor_all_all_adverse_event_select_features_symbol_all.tsv) were mapped to [associated gene-disease connections](target_disease/descriptor_all_all_adverse_event_select_features_symbol_connection_map.tsv). [Enrichment analysis](target_disease/descriptor_all_all_adverse_event_select_features_symbol_connection_enrich.tsv) was performed to test whether the selected targets are enriched for disease-related genes.

## Functional enrichment of selected target genes 

+ Functional enrichment analysis was performed for [15 adverse events of interest](target_enrichment/adverse_event_of_interest.txt). Four types of functional annotations were included:
  + [Reactome pathways](target_enrichment/descriptor_all_all_adverse_event_select_features_reactome_enrichment_results.tsv)
  + [Biological Process (BP) of Gene Ontology](target_enrichment/descriptor_all_all_adverse_event_select_features_go_bp_enrichment_results.tsv)
  + [Molecular Function (MF) of Gene Ontology](target_enrichment/descriptor_all_all_adverse_event_select_features_go_mf_enrichment_results.tsv)
  + [Cellular Component (CC) of Gene Ontology](target_enrichment/descriptor_all_all_adverse_event_select_features_go_cc_enrichment_results.tsv)

## Predictive target map of adverse events 
 
+ [Mapping between 36 adverse events and 101 targets](target_map/descriptor_all_all_adverse_event_select_features_data.tsv) were generated. Adverse events with AUROC > 0.65 were included in the map. Targets that are mapped to more than 5 adverse events are included in the map. 
